nodego/events.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nodego
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"net/http"
+	"runtime/debug"
+	"time"
+
+	"google.golang.org/api/pubsub/v1"
+	"google.golang.org/api/storage/v1"
+)
+
+// JSTime is a wrapper for time.Time to decode the time from Javascript.
+type JSTime struct {
+	time.Time
+}
+
+// UnmarshalJSON parses a JSON string to time.Time.
+func (t *JSTime) UnmarshalJSON(b []byte) (err error) {
+	if string(b) == `"null"` {
+		t.Time = time.Time{}
+		return
+	}
+
+	t.Time, err = time.Parse(`"2006-01-02T15:04:05.000Z"`, string(b))
+	return
+}
+
+// EventContext holds the data associated with the event that triggered the
+// execution of the function along with metadata of the event itself.
+type EventContext struct {
+	EventID   string `json:"eventId"`
+	Timestamp JSTime `json:"timestamp"`
+	EventType string `json:"eventType"`
+	Resource  string `json:"resource"`
+}
+
+// FunctionEvent is the basic data structure passed to functions by non-HTTP
+// triggers.
+type FunctionEvent struct {
+	Context EventContext
+	Data    json.RawMessage
+}
+
+// UnmarshalJSON parses a JSON string to time.Time.
+func (e *FunctionEvent) UnmarshalJSON(b []byte) error {
+	raws := map[string]json.RawMessage{}
+	if err := json.Unmarshal(b, &raws); err != nil {
+		return err
+	}
+
+	rawContext, ok := raws["context"]
+	if !ok {
+		rawContext = b
```

Comment:
Does this case actually happen in practice? The documentation that you linked to seems to indicate that this won't happen: https://cloud.google.com/functions/docs/writing/background

Reply:
This is a fallback for old behaviour for pub sub triggers.

For the time being, this new behaviour is hidden behind a gcloud config flag (`functions/use_new_pubsub_trigger`)

The full message from `gcloud functions deploy` when you use `--trigger-topic` is:

```
WARNING: The default event schema returned by Cloud Functions with a pub/sub (`--trigger-topic`) trigger will soon change. The `eventId`, `timestamp`, `eventType`, and `resource` properties will all be moved into the `event.context` property. To opt-in to the new behavior early, run`gcloud config set functions/use_new_pubsub_trigger True`. To restore old behavior you can run `gcloud config set functions/use_new_pubsub_trigger False` or use the `--trigger-event` flag e.g. `gcloud functions deploy --trigger-event providers/cloud.pubsub/eventTypes/topic.publish --trigger-resource <TOPIC_NAME>` Please see https://cloud.google.com/functions/docs/writing/background#event_parameterfor more information on the new schema.
```

Reply:
Oh, ok. Thanks for the explanation.